### PR TITLE
Performance: Adds optimized request headers

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

This uses a generated file to create a list of well known headers for Item point operations. This allows the properties to be directly set instead of doing a dictionary lookup.

This makes CosmosMessageHeadersInternal a contract which then has custom implementation for request headers and response headers.

Overall +50 to 100 Rps improvement for read item stream operations

*Create is just creating the object
*Read is object creation + adding all the headers used for a normal read scenario
*Keys is used in transport layer to iterate and populate the transport objects
image

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Headers/CosmosMessageHeadersInternal.cs [hunk 1]
```diff
@@ -9,167 +9,207 @@ namespace Microsoft.Azure.Cosmos
...
-    using System.Linq;
+    using System.Text;
+    using Microsoft.Azure.Documents;
     using Microsoft.Azure.Documents.Collections;
 
-    /// <summary>
-    /// Internal header class with priority access for known headers and support for dictionary-based access to other headers.
-    /// </summary>
-    internal class CosmosMessageHeadersInternal : INameValueCollection
+    internal abstract class CosmosMessageHeadersInternal : IEnumerable, INameValueCollection
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Headers/CosmosMessageHeadersInternal.cs:16]: INameValueCollection already has IEnumerable

File: Microsoft.Azure.Cosmos/src/Headers/Headers.cs [hunk 2]
```diff
@@ -148,65 +141,76 @@ internal string SubStatusCodeLiteral
...
+            this.CosmosMessageHeaders = new StoreRequestNameValueCollection();
         }
 
-        internal Headers(INameValueCollection nameValue)
+        internal Headers(INameValueCollection nameValueCollection)
         {
-            this.CosmosMessageHeaders = nameValue;
+            if (nameValueCollection is StoreResponseNameValueCollection storeResponseNameValueCollection)
+            {
+                this.CosmosMessageHeaders = new StoreResponseHeaders(storeResponseNameValueCollection);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Headers/Headers.cs:210]: should it return? Otherwise it's setting `this.CosmosMessageHeaders` again below

**Review by ealsur — APPROVED** [review]
LGTM

**Comment by kirankumarkolli** [comment]
What's the E2E impact? 

From the attached chart reads are more optimal compared to creates. Does that fit well into the overall requirements?

**Review by kirankumarkolli — APPROVED** [review]

**Comment by j82w** [comment]
> What's the E2E impact?
> 
> From the attached chart reads are more optimal compared to creates. Does that fit well into the overall requirements?

Please see the updated PR description.

**Pull request merged**